miniapps/parelag/MultilevelHcurlSolver.cpp

```diff
@@ -0,0 +1,739 @@
+                                            vector<int> {1});
+
+   // A list of (piecewise) constant values for the coefficient 'alpha', in
+   // accordance with the mesh attributes. If only a single entry is given, it
+   // is applied to the whole mesh/domain.
+   vector<double> alpha_vals = prob_list.Get("alpha values",
+                                             vector<double> {1.0});
+
+   // A list of (piecewise) constant values for the coefficient 'beta', in
+   // accordance with the mesh attributes. If only a single entry is given, it
+   // is applied to the whole mesh/domain.
+   vector<double> beta_vals = prob_list.Get("beta values",
+                                            vector<double> {1.0});
+
+   // The list of solvers to invoke.
+   auto list_of_solvers = prob_list.Get<list<string>>("List of linear solvers");
+
+   ostringstream mesh_msg;
+   if (!myid)
+   {
+      mesh_msg << '\n' << string(50, '*') << '\n'
+               << "*                     Mesh: " << meshfile << "\n*\n"
+               << "*                 FE order: " << feorder << '\n'
+               << "*          Upscaling order: " << upscalingOrder << "\n*\n";
+   }
+
+   // Read the (serial) mesh from the given mesh file and uniformly refine it.
+   shared_ptr<ParMesh> pmesh;
+   {
+      if (!myid)
+      {
+         cout << "\nReading and refining serial mesh...\n";
+         cout << "Times to refine mesh in serial: " << ser_ref_levels << ".\n";
+      }
+
+      ifstream imesh(meshfile);
+      if (!imesh)
+      {
+         if (!myid)
+         {
+            cerr << "ERROR: Cannot open mesh file: " << meshfile << ".\n";
+         }
+         return EXIT_FAILURE;
+      }
+
+      auto mesh = make_unique<Mesh>(imesh, true, true);
+      imesh.close();
+
+      for (int l = 0; l < ser_ref_levels; ++l)
+      {
+         if (!myid)
+         {
+            cout << "Serially refining mesh: " << l + 1 << "...\n";
+         }
+         mesh->UniformRefinement();
+      }
+
+      if (ser_ref_levels < 0)
+      {
+         ser_ref_levels = 0;
+         for (; mesh->GetNE() < 6 * num_ranks; ++ser_ref_levels)
+         {
+            if (!myid)
+            {
+               cout << "Serially refining mesh: " << ser_ref_levels + 1
+                    << "...\n";
+            }
+            mesh->UniformRefinement();
+         }
+      }
+
+      if (!myid)
+      {
+         cout << "Times refined mesh in serial: " << ser_ref_levels << ".\n";
+         cout << "Building and refining parallel mesh...\n";
+         cout << "Times to refine mesh in parallel: " << par_ref_levels
+              << ".\n";
+         mesh_msg << "*    Serial refinements: " << ser_ref_levels << '\n'
+                  << "*      Coarse mesh size: " << mesh->GetNE() << "\n*\n";
+      }
+
+      pmesh = make_shared<ParMesh>(comm, *mesh);
+   }
+
+   // Mark essential boundary attributes.
+   MFEM_ASSERT(par_ess_attr.size() <= 1 ||
```

Comment:
Since `MFEM_ASSERT` is checked only in debug mode, I would prefer to use `MFEM_VERIFY` in all instances in this file. When verifying input like this, the cost of the check is negligible, and checking in optimized mode can help the user find input errors.